proposals/extensions_v2.md

```diff
@@ -0,0 +1,368 @@
+    {
+        // Original body
+    }
+}
+
+extension Enumerable
+{
+    public TResult Sum<TSource, TResult, TAccumulator>(Func<TSource, TResult> selector)
+        for IEnumerable<TSource> source
+        where TResult : struct, INumber<TResult>
+        where TAccumulator : struct, INumber<TAccumulator>
+    {
+        // Exactly the same code as original body.
+    }
+}
+```
+
+This form supports *non* extension static methods.  For example `Enumerable.Range` would migrate like so:
+
+```c#
+static class Enumerable
+{
+    public static IEnumerable<int> Range(int start, int count) { ... }
+}
+
+
+extension Enumerable
+{
+    // Exact same signature.  No 'for-clause'.
+    public static IEnumerable<int> Range(int start, int count) { ... }
+}
+```
+
+## Disambiguation
+
+Classic extension methods today can be disambiguated by falling back to static-invocation syntax.  For example, if `x.Count()` is ambiguous, it is possible to switch to some form of `StaticClass.Count(x)` to call the desired method.  A similar facility is needed for modern extension methods.  While the existing method-invocation-translation approach works fine for methods (where the receiver can be remapped to the first argument of the static extension method call), it is ungainly for these other extension forms.
+
+As an initial strawman this proposal suggests reusing `cast expression` syntax for disambiguation purposes.  For example:
+
+```c#
+var v1 = ((Extension)receiver).ExtensionMethod(); // instead of Extension.ExtensionMethod(receiver)
+var v2 = ((Extension)receiver).ExtensionProperty;
+var v3 = ((Extension)receiver)[indexerArg];
+var v4 = (Extension)receiver1 + receiver2;
+```
+
+Constructors and static methods would not need any special syntax as the extension can cleanly be referenced as a type where needed.
+
+```c#
+var v3 = new Extension(...); // Makes instance of the actual extended type.
+var v4 = Extension.StaticExtensionMethod(...);
+```
+
+Note 1: while the cast syntax traditionally casts or converts a value, that would not be the case for its use here.  It would only be used as a lookup mechanism to indicate which extension gets priority.  Importantly, even with this syntax, extensions themselves are not types.  For example:
+
+```c#
+Extension e1;                   // Not legal.  Extension is not a type.
+Extension[] e2;                 // Not legal.  Extension is not a type.
+List<Extension> e3;             // Not legal.  Extension is not a type.
+var v1 = (Extension)receiver;   // Not legal.  Can't can't have a value of extension type.
+```
+
+This is exactly the same as the restrictions on static-types *except* with the carve out that you can use the extension in a cast-syntax or new-expression *only* for lookup purposes, or where a static-class could be used, and nothing else.  Usage in places like `nameof(Extension)` or `typeof(Extension)` would still be fine, as those are places where a static type is allowed.
```

Comment:
@jnm2 help me wordsmith :D

Reply:
Looks fine to me!

I could imagine confusion that `Activator.CreateInstance(typeof(Extension))` and `new Extension()` mean different things, so I'm weighing the use cases of even having `typeof` and `nameof`.

Reply:
we need typeof/nameof for perfect compat.

Reply:
any place a static-class can be used, you must be able to use an extensino.  or else we break compat :)